File: week3/c++/1LongestSubstringWithoutRepeatingCharacters.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> memo;
        int res = 0, i = 0, j = 0;
        while (j < s.length()) {
            if (memo.find(s[j]) == memo.end() || memo[s[j]] < i) {
                memo[s[j]] = j;
                res = max(j - i + 1, res);
            }
            else {
                i = memo[s[j]] + 1;
                memo[s[j]] = j;
            }
            j++;
        }
        return res;
    }
};

File: week3/Java/1LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> memo = new HashMap<>();
        int i = 0, j = 0;
        int res = 0;
        while (j < s.length()) {
            if (!memo.containsKey(s.charAt(j)) || memo.get(s.charAt(j)) < i) {
                memo.put(s.charAt(j), j);
            }
            else {
                i = memo.get(s.charAt(j)) + 1;
                memo.put(s.charAt(j), j);
            }
            res = Math.max(res, j - i + 1);
            j++;
        }
        return res;
    }
}